### DISCUSSION TITLE
Horizontal sharding with shared tables

### DISCUSSION THREAD

**Discussion by MatejHamala** [sqlalchemy/sqlalchemy#11101] [Usage Questions] [answered]
I am using horizontal sharding extension in a setup where some tables are shared across all databases. Those shared tables are stored in the main database but accessible from shards (using MariaDB's SPIDER :spider:  engine). However, the possibility of accessing the same data from multiple different shards creates an issue within SQLAlchemy with duplicate ORM objects in sessions. In the event that I retrieve the same row from a shared table from two different shards, the session creates two different identity keys for them due to the `shard_id` being part of the identity key, set using the `identity_token` by `ShardedSession`. Therefore, I would like to inquire if you have any ideas on how to solve this problem. :slightly_smiling_face:

The least invasive approach that occurred to me is to "decorate" the methods for generating identity keys of `mapper`s (`identity_key_from_primary_key` and `identity_key_from_row`) for shared tables so that they replace the passed `identity_token` with the `shard_id` of the main database, for example, as follows:
```python
@event.listens_for(Base, 'mapper_configured', propagate=True)
def receive_mapper_configured(mapper, class_):
    if not class_.__table__.is_shared:
        return

    def with_replaced_identity_token(fn):
        @wraps(fn)
        def wrapper(*args, identity_token=None, **kwargs):
            return fn(*args, identity_token=MAIN_DB_ID, **kwargs)

        return wrapper

    mapper.identity_key_from_primary_key = with_replaced_identity_token(
        mapper.identity_key_from_primary_key
    )
    mapper.identity_key_from_row = with_replaced_identity_token(
        mapper.identity_key_from_row
    )
```
However, this solution doesn't work :disappointed: because the creation of identity keys isn't unified into these methods, instead, it's hard-coded in several places in the [loading.py](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/loading.py) module, as follows:
https://github.com/sqlalchemy/sqlalchemy/blob/b5e8c80f24c326e8491d1db6c5b32dcf730b53b4/lib/sqlalchemy/orm/loading.py#L1067-L1071
https://github.com/sqlalchemy/sqlalchemy/blob/b5e8c80f24c326e8491d1db6c5b32dcf730b53b4/lib/sqlalchemy/orm/loading.py#L1218-L1222
I don't know if there's any reason for it, but if not, would it be possible to unify it here? :blush: Is there any other better place where I could change the value of `identity_token` for a specific ORM classes?

For completeness, I'll add that `identity_chooser` always returns the id of the main database for shared tables. So, this problem doesn't arise when fetching individual rows from these tables. However, if this row is loaded together with rows from purely sharded tables, for example, using lazy / joined loading, `identity_chooser` is not called, and if I understand correctly, the `identity_token` is inherited from the parent object / determined by `execute_chooser`.

Thank you in advance for any ideas and assistance! :wink:

**Comment by zzzeek** [comment]
> However, the possibility of accessing the same data from multiple different shards creates an issue within SQLAlchemy with duplicate ORM objects in sessions. In the event that I retrieve the same row from a shared table from two different shards, the session creates two different identity keys for them due to the shard_id being part of the identity key, set using the identity_token by ShardedSession. Therefore, I would like to inquire if you have any ideas on how to solve this problem. 🙂

in a sharded system defined as the way the "horizontal shard" recipe sees it, the different shards don't have "the same data".  each object identity is unique to its shard.    if you have a system where the same object identity is in multiple backends, that sounds more like **replication**, not shards.       in any case, the main thing the sharding extension gives you is this ability to pull objects from different databases, which have the same primary key identity but different source DB therefore are different identities.  if two rows pulled from database A and database B with the same primary key identity should in fact be treated as the same object, then this is not appropriate for the horizontal sharding API.

overall horizontal shard is delivered "as is" and isn't really known to have been successful at what it claims, usually a particular application will need to find its own way of solving the distribution problems it has either with middleware or specific client side programming patterns.  I would design those first, then consider how well or not the sharding extension works in that context.

**Comment by MatejHamala** [comment]
I agree that shared tables don't quite fit into the horizontal sharding schema, but they are only a specific side matter complementing fully sharded tables with a large amount of data that need to be distributed because one database couldn't handle them alone. Apart from this single issue (which can be worked around), the solution using horizontal sharding in SQLAlchemy works very well. However, I understand that this is not a standard use case and that this tool is not adapted to it.

However, I would like to repeat the question about a unified way of creating identity keys. Is it really correct like this? According to the [documentation](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.Mapper.identity_key_from_row):
> Return an identity-map key for use in storing/retrieving an item from an identity map.

I would expect SQLAlchemy to use methods available in the mapper of the given class to create identity keys, rather than creating them manually (especially when the mapper instance is available in these places). Is there a chance that this will be unified in the future? :slightly_smiling_face:

**Comment by zzzeek** [comment]
> 
> > Return an identity-map key for use in storing/retrieving an item from an identity map.
> 
> I would expect SQLAlchemy to use methods available in the mapper of the given class to create identity keys, rather than creating them manually (especially when the mapper instance is available in these places). Is there a chance that this will be unified in the future? 🙂


no, that's a public method but has no implication on the internals and what you're doing with the monkeypatching there is not recommended and should not be expected to do anything.

SQLAlchemy adds features based on use cases that are illustrated up front and for which it's agreed there's a need to add that API.

in this case, the identity token originates from the horizontal sharding extension, which really is a glorified example class.  it really shouldn't be an extension.   Take that module and modifiy it to your needs, removing the identity token logic that it adds as needed and/or make it conditional based on your classes.

**Comment by MatejHamala** [comment, marked as answer]
Alright, modifying the module also seems to be a viable path without significant intervention into SQLAlchemy internals. In the end, it's just a few adjustments that seem to be functional. Thank you once again for your response!
